README.md
# Webfinger

TODO: Write a gem description

## Installation

Add this line to your application's Gemfile:

    gem 'webfinger'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install webfinger

## Usage

TODO: Write usage instructions here

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## [Unreleased]

## [2.1.1] - 2022-10-09

### Changed

- convert Faraday::Error to WebFinger::Exception https://github.com/nov/webfinger/commit/a0d8da20f6ce819eea8c2d2c81bcc90d3f41df6f

## [2.1.1] - 2022-10-08

### Fixed

- make response json indifferent accessable https://github.com/nov/webfinger/commit/806bc2e2220024f419f44857e5629ff7f9f78fbf

## [2.1.0] - 2022-10-08

### Changed

- Moved to faraday v2 https://github.com/nov/webfinger/commit/e72fa7b7c7018c809ba9ce8aae2ee64d6c1d1d02

## [2.0.0] - 2022-10-08

### Added

- start recording CHANGELOG

### Changed

- Switch from httpclient to faraday https://github.com/nov/webfinger/pull/9

spec/spec_helper.rb
require 'simplecov'

SimpleCov.start do
  add_filter 'spec'
end

require 'rspec'
require 'webfinger'
require 'helpers/webmock_helper'

spec/webfinger_spec.rb
require 'spec_helper'

describe WebFinger do
  let(:resource) { 'acct:nov@example.com' }

  describe '#discover!' do
    {
      'example.com'             => 'https://example.com',
      'example.com/~nov/'       => 'https://example.com',
      'nov@example.com'         => 'https://example.com',
      'nov.matake@example.com'  => 'https://example.com',
      'acct:nov@example.com'    => 'https://example.com',
      'mailto:nov@example.com'  => 'https://example.com',
      'device:example.com'      => 'https://example.com',
      'unknown:nov@example.com' => 'https://example.com',
      'http://example.com/nov'  => 'https://example.com',
      'https://example.com/nov' => 'https://example.com',
      'example.com:8080'             => 'https://example.com:8080',
      'example.com:8080/~nov/'       => 'https://example.com:8080',
      'nov@example.com:8080'         => 'https://example.com:8080',
      'nov.matake@example.com:8080'  => 'https://example.com:8080',
      'acct:nov@example.com:8080'    => 'https://example.com:8080',
      'mailto:nov@example.com:8080'  => 'https://example.com:8080',
      'device:example.com:8080'      => 'https://example.com:8080',
      'unknown:nov@example.com:8080' => 'https://example.com:8080',
      'http://example.com:8080'      => 'https://example.com:8080',
      'https://example.com:8080'     => 'https://example.com:8080',
      'discover.example.com'             => 'https://discover.example.com',
      'discover.example.com/~nov/'       => 'https://discover.example.com',
      'nov@discover.example.com'         => 'https://discover.example.com',
      'nov.matake@discover.example.com'  => 'https://discover.example.com',
      'acct:nov@discover.example.com'    => 'https://discover.example.com',
      'mailto:nov@discover.example.com'  => 'https://discover.example.com',
      'device:discover.example.com'      => 'https://discover.example.com',
      'unknown:nov@discover.example.com' => 'https://discover.example.com',
      'http://discover.example.com/nov'  => 'https://discover.example.com',
      'https://discover.example.com/nov' => 'https://discover.example.com',
      'discover.example.com:8080'             => 'https://discover.example.com:8080',
      'discover.example.com:8080/~nov/'       => 'https://discover.example.com:8080',
      'nov@discover.example.com:8080'         => 'https://discover.example.com:8080',
      'nov.matake@discover.example.com:8080'  => 'https://discover.example.com:8080',
      'acct:nov@discover.example.com:8080'    => 'https://discover.example.com:8080',
      'mailto:nov@discover.example.com:8080'  => 'https://discover.example.com:8080',
      'device:discover.example.com:8080'      => 'https://discover.example.com:8080',
      'unknown:nov@discover.example.com:8080' => 'https://discover.example.com:8080',
      'http://discover.example.com:8080/nov'  => 'https://discover.example.com:8080',
      'https://discover.example.com:8080/nov' => 'https://discover.example.com:8080'
    }.each do |resource, base_url|
      endpoint = File.join base_url, '/.well-known/webfinger'
      context "when resource=#{resource}" do
        it "should access to #{endpoint}" do
          mock_json endpoint, 'all', query: {resource: resource} do
            response = WebFinger.discover! resource
            response.should be_instance_of WebFinger::Response
          end
        end
      end
    end

    context 'with rel option' do
      shared_examples_for :discovery_with_rel do
        let(:query_string) do
          query_params = [{resource: resource}.to_query]
          Array(rel).each do |_rel_|
            query_params << {rel: _rel_}.to_query
          end
          query_params.join('&')
        end

        it 'should request with rel' do
          query_string.scan('rel').count.should == Array(rel).count
          mock_json 'https://example.com/.well-known/webfinger', 'all', query: query_string do
            response = WebFinger.discover! resource, rel: rel
            response.should be_instance_of WebFinger::Response
          end
        end
      end

      context 'when single rel' do
        let(:rel) { 'http://openid.net/specs/connect/1.0/issuer' }
        it_behaves_like :discovery_with_rel
      end

      context 'when multiple rel' do
        let(:rel) { ['http://openid.net/specs/connect/1.0/issuer', 'vcard'] }
        it_behaves_like :discovery_with_rel
      end
    end

    context 'when error' do
      {
        400 => WebFinger::BadRequest,
        401 => WebFinger::Unauthorized,
        403 => WebFinger::Forbidden,
        404 => WebFinger::NotFound,
        500 => WebFinger::HttpError
      }.each do |status, exception_class|
        context "when status=#{status}" do
          it "should raise #{exception_class}" do
            expect do
              mock_json 'https://example.com/.well-known/webfinger', 'all', query: {resource: resource}, status: [status, 'HTTPError'] do
                response = WebFinger.discover! resource
              end
            end.to raise_error exception_class
          end
        end
      end
    end
  end

  describe '#logger' do
    subject { WebFinger.logger }

    context 'as default' do
      it { should be_instance_of Logger }
    end

    context 'when specified' do
      let(:logger) { 'Rails.logger or something' }
      before { WebFinger.logger = logger }
      it { should == logger }
    end
  end

  describe '#debugging?' do
    subject { WebFinger.debugging? }

    context 'as default' do
      it { should == false }
    end

    context 'when debugging' do
      before { WebFinger.debug! }
      it { should == true }

      context 'when debugging mode canceled' do
        before { WebFinger.debugging = false }
        it { should == false }
      end
    end
  end

  describe '#url_builder' do
    subject { WebFinger.url_builder }

    context 'as default' do
      it { should == URI::HTTPS }
    end

    context 'when specified' do
      let(:url_builder) { 'URI::HTTP or something' }
      before { WebFinger.url_builder = url_builder }
      it { should == url_builder }
    end
  end

  describe '#http_client' do
    subject { WebFinger.http_client }

    describe '#logger' do
      subject { WebFinger.http_client.builder.handlers.collect(&:klass) }

      context 'as default' do
        it { should_not include Faraday::Response::Logger }
      end

      context 'when debugging' do
        before { WebFinger.debug! }
        it { should include Faraday::Response::Logger }
      end
    end
  end
end


spec/helpers/webmock_helper.rb
require 'webmock/rspec'

module WebMockHelper
  def mock_json(endpoint, response_file, options = {})
    endpoint = endpoint.to_s
    stub_request(:get, endpoint).with(
      request_for(options)
    ).to_return(
      response_for(response_file, options)
    )
    yield
    a_request(:get, endpoint).with(
      request_for(options)
    ).should have_been_made.once
  end

  private

  def request_for(options = {})
    request = {}
    if options[:query]
      request[:query] = options[:query]
    end
    request
  end

  def response_for(response_file, options = {})
    response = {}
    response[:headers] = {
      'Content-Type': 'application/json'
    }
    response[:body] = File.new(File.join(File.dirname(__FILE__), '../mock_json', "#{response_file}.json"))
    if options[:status]
      response[:status] = options[:status]
    end
    response
  end
end

include WebMockHelper
WebMock.disable_net_connect!


lib/webfinger.rb
require 'json'
require 'faraday'
require 'faraday/follow_redirects'
require 'active_support'
require 'active_support/core_ext'

module WebFinger
  VERSION = File.read(
    File.join(File.dirname(__FILE__), '../VERSION')
  ).delete("\n\r")

  module_function

  def discover!(resource, options = {})
    Request.new(resource, options).discover!
  end

  def logger
    @logger
  end
  def logger=(logger)
    @logger = logger
  end
  self.logger = ::Logger.new(STDOUT)
  logger.progname = 'WebFinger'

  def debugging?
    @debugging
  end
  def debugging=(boolean)
    @debugging = boolean
  end
  def debug!
    self.debugging = true
  end
  self.debugging = false

  def url_builder
    @url_builder ||= URI::HTTPS
  end
  def url_builder=(builder)
    @url_builder = builder
  end

  def http_client
    Faraday.new(headers: {user_agent: "WebFinger #{VERSION}"}) do |faraday|
      faraday.response :raise_error
      faraday.response :json
      faraday.response :follow_redirects
      faraday.response :logger, WebFinger.logger if debugging?
      faraday.adapter Faraday.default_adapter
      http_config.try(:call, faraday)
    end
  end

  def http_config(&block)
    @http_config ||= block
  end
end

require 'webfinger/exception'
require 'webfinger/request'
require 'webfinger/response'


lib/webfinger/request.rb
module WebFinger
  class Request
    attr_accessor :resource, :relations

    def initialize(resource, options = {})
      self.resource = resource
      self.relations = Array(options[:rel] || options[:relations])
      @options = options
    end

    def discover!
      handle_response do
        WebFinger.http_client.get(endpoint.to_s)
      end
    end

    private

    def endpoint
      path = '/.well-known/webfinger'
      host, port = host_and_port
      WebFinger.url_builder.build [nil, host, port, path, query_string, nil]
    end

    def host_and_port
      uri = URI.parse(resource) rescue nil
      if uri.try(:host).present?
        [uri.host, [80, 443].include?(uri.port) ? nil : uri.port]
      else
        scheme_or_host, host_or_port, port_or_nil = resource.split('@').last.split('/').first.split(':')
        case host_or_port
        when nil, /\d+/
          [scheme_or_host, host_or_port.try(:to_i)]
        else
          [host_or_port, port_or_nil.try(:to_i)]
        end
      end
    end

    def query_string
      query_params.join('&')
    end

    def query_params
      query_params = [{resource: resource}]
      relations.each do |relation|
        query_params << {rel: relation}
      end
      query_params.collect(&:to_query)
    end

    def handle_response
      json = yield.body.with_indifferent_access
      Response.new json
    rescue Faraday::Error => e
      case e.response_status
      when nil
        raise Exception.new e
      when 400
        raise BadRequest.new('Bad Request', e.response_body)
      when 401
        raise Unauthorized.new('Unauthorized', e.response_body)
      when 403
        raise Forbidden.new('Forbidden', e.response_body)
      when 404
        raise NotFound.new('Not Found', e.response_body)
      else
        raise HttpError.new(e.response_status, e.response_body, e.response_body)
      end
    end
  end
end
